### ISSUE TITLE
Complex logical expression triggers poor performance

### ISSUE THREAD

**Issue by edsrzf** [vimeo/psalm#9565]
This 40-line example takes 5-10 seconds to analyze: https://psalm.dev/r/f171a48a60

This example is an anonymized version of real code that was causing a particular file in a code base I'm working in to take 10-15 seconds to analyze. It's not very good code, but it shouldn't take this long to analyze.

A lot of the time is spent inside `Algebra::simplifyCNF`, where it handles 4097 clauses. I don't fully understand what the purposes of this function is yet so I don't know whether or not that's reasonable, but it seems like a lot. It also gets called 3 separate times with that many clauses, so may be duplicating work.

**Comment by orklah** [comment]
`Algebra::simplifyCNF` is the magic behind Psalm handling of conditions.

Among a lot of other things, It's what allows it to make some deductions like that: https://psalm.dev/r/f171a48a60

But yeah, for enormous conditions like yours, the sheer amount of clauses it has to generate is enough to make it process the file for some time.

We already have somewhat arbitrary limits to the number of clauses we allow it to process for example: https://github.com/vimeo/psalm/blob/3c4a829f712da775ca1b0ec2705622e3c2ff61a1/src/Psalm/Internal/Algebra.php#L83

But it's a tradeoff between time and Psalm just giving up on some checks.

**Comment by edsrzf** [comment]
It looks like you may have accidentally re-pasted my original snippet instead of a new one. I think I roughly understand the purpose of this code though.

Initially I didn't understand what Psalm could deduce from logical or (`||`). Now I've realized that it can't deduce much when the entire condition is true, but if it's false, it could realize that all of the input conditions (and thus variables) are also false.

It appears that Psalm performs these simplifications eagerly, even if the variables involved are never used inside the conditionals. It seems like there could be an opportunity to do it lazily, when the variable types are actually needed. This seems like a pretty large change though.

**Comment by orklah** [comment]
Right, this is what I wanted to post: https://psalm.dev/r/f59fabea8f
